codegen/ace/src/codegen.rs

```diff
@@ -0,0 +1,257 @@
+
+        // The polynomials are stored in the circuits as constants
+        let periodic_polys: BTreeMap<QualifiedIdentifier, Vec<NodeIndex>> = periodic_polys
+            .iter()
+            .map(|(ident, poly)| {
+                let poly = poly
+                    .iter()
+                    .map(|c| circ.constant(*c))
+                    .collect::<Vec<NodeIndex>>();
+                (*ident, poly)
+            })
+            .collect();
+
+        let periodic_values: Vec<(NodeIndex, PeriodicColumnAccess)> = periodic_ops
+            .iter()
+            .map(|(idx, op)| match op {
+                Operation::Value(Value::PeriodicColumn(pca)) => (*idx, *pca),
+                _ => unreachable!("periodic values already filtered"),
+            })
+            .collect();
+
+        let z_min_num_cycles = circ.input();
+
+        let max_cycle_len = periodic_values
+            .iter()
+            .map(|(_, pca)| pca.cycle)
+            .max()
+            .unwrap(); //TODO proper error
+
+        // Map from cycle_len to k = max_cycle_len / cycle_len, this step
+        // also removes duplicated cycle_len
+        let mut ks: BTreeMap<usize, usize> = BTreeMap::new();
+        periodic_values.iter().for_each(|(_, pca)| {
+            let k = max_cycle_len / pca.cycle;
+            ks.insert(pca.cycle, k);
+        });
+
+        // Map from cycle_len to z^l
+        let z_ls: BTreeMap<usize, NodeIndex> = ks
+            .into_iter()
+            .map(|(cycle_len, k)| {
+                let k = u32::try_from(k).unwrap();
+                (cycle_len, circ.pow(z_min_num_cycles, k))
+            })
+            .collect();
+
+        periodic_values.iter().for_each(|(idx, pca)| {
+            let z_l = z_ls[&pca.cycle];
+            let poly = &periodic_polys[&pca.name];
+            let new_idx = circ.horners_method(z_l, poly);
+            mapping.insert(*idx, new_idx);
+        });
+    }
+}
+
+/// Air constraints are separated in 3 main groups: integrity roots,
+/// boundary first roots and boundary last roots (bl).
+/// The roots in each group are linearly combined with powers of a random challenge `alpha`:
+/// - `int = \sum_i int_roots[i] * alpha^i` with `i` from 0 to |int_roots|
+/// - `bf = \sum_i bf_roots[i] * alpha^{i+offset}` with `i` from 0 to |int_roots| and offset = |int_roots|
+/// - `bl = \sum_i bl_roots[i] * alpha^{i+offset}` with `i` from 0 to |bf_roots| and offset = |int_roots| + |bf_roots|
+///
+/// The 3 groups are joined in the formula:
+/// `int * z_g / zn_1 + bf / z_1 + bl / z_g`
+/// where
+/// - `z_g = z - g^(-1)`
+/// - `z_1 = z - 1`
+/// - `zn_1 = z^n -1`
+///
+/// Which can be rewritten as
+/// `int * z_g^2 * z_1 + bf * zn_1 * z_g + bl * zn_1 * z_1 / (zn_1 * z_1 * z_g)`
+/// This function returns a circuit computing the numerator above, given the following inputs:
+/// - alpha
+/// - z
+/// - z^n
+/// - g^(-1)
+pub fn build_ace_circuit(ir: &Air) -> anyhow::Result<(NodeIndex, Circuit)> {
```

Comment:
nit: this seems to be the main function in this module - I'd move it to the top of the file. The other functions/structs could go into helper sections after it.